src/Extensions.Hosting.AsyncInitialization/Hosting/AsyncInitializationHostExtensions.cs

```diff
@@ -101,26 +101,40 @@ public static async Task InitAndRunAsync(this IHost host, TimeSpan teardownTimeo
             await using (host.AsAsyncDisposable().ConfigureAwait(false))
             {
                 cancellationToken.ThrowIfCancellationRequested();
+                
+                using var cts = new CancellationTokenSource();
+                Exception? innerException = null;
                 try
                 {
-                    await host.InitAsync(cancellationToken).ConfigureAwait(false);
-                    await host.StartAsync(cancellationToken).ConfigureAwait(false);
-                    await host.WaitForShutdownAsync(cancellationToken).ConfigureAwait(false);
-                }
-                finally
-                {
-                    using var cts = new CancellationTokenSource(teardownTimeout);
                     try
                     {
+                        await host.InitAsync(cancellationToken).ConfigureAwait(false);
+                        await host.StartAsync(cancellationToken).ConfigureAwait(false);
+                        await host.WaitForShutdownAsync(cancellationToken).ConfigureAwait(false);
+                    }
+                    catch (Exception ex)
+                    {
+                        innerException = ex;
+                        throw;
+                    }
+                    finally
+                    {
+                        cts.CancelAfter(teardownTimeout);
                         await host.TeardownAsync(cts.Token).WaitAsync(cts.Token).ConfigureAwait(false);
                     }
-                    catch (OperationCanceledException) when (cts.IsCancellationRequested)
+                }
+                catch (OperationCanceledException) when (cts.IsCancellationRequested)
+                {
+                    var timeoutException = new TimeoutException("Teardown cancelled due to timeout.");
+                    if (innerException != null)
                     {
-                        throw new TimeoutException("Teardown cancelled due to timeout.");
+                        throw new AggregateException(timeoutException, innerException);
```

Comment:
nit: I would put `innerException` first, since it happened first

Reply:
Done